### ISSUE TITLE
Add miniPCIe interface to the board design?

### ISSUE THREAD

**Issue by danielkucera** [lora-gateway/esxp1302#4]
Hello,

I am successfully using this firmware with [LR1302 from Elecrow](https://www.elecrow.com/lr1302-lorawan-gateway-module-spi-eu868-sx1302-long-range-gateway-module-support-8-channels.html) and my [miniPCIe breakout board](https://lectronz.com/products/minipci-e-breakout-board).
This is my current setup.

![IMG20231217152824](https://github.com/lora-gateway/esxp1302/assets/1734256/a637d0cd-d4ff-4bd9-8c4b-d3732de7dabf)

Would you consider adding miniPCIe connector to your current HW design according to [this spec](https://www.elecrow.com/wiki/index.php?title=LR1302_LoRaWAN_Gateway_Module#Interface_Function)?
I tested the same setup also for [RAK2287](https://docs.rakwireless.com/Product-Categories/WisLink/RAK2287/Datasheet/#pinout-description) and it works .

I was considering doing it myself but I don't have the design software to edit your project and I am lazy to design it from scratch.

**Comment by lora-gateway** [comment]
@danielkucera:
Thanks for this proposal. Since the market already has some miniPCIe based modules available, it would be easier for people to have the hardware ready to play with.

First, I'd like to check if all of them (or most of them) follow the same spec; then I plan to create an interface board to bridge the miniPCIe connector to the motherboard. I need help from a friend to do this, so it would take some time.

BTW: What's your design software? Is it KiCAD? Maybe in the future, someone can help to convert the hardware design to such tool so more people can modify directly...

**Comment by danielkucera** [comment]
Thanks for the reply. Yes, I prefer KiCad. I even managed to import the board design and I am trying to add the interface.
This is my current WiP:
![image](https://github.com/lora-gateway/esxp1302/assets/1734256/f645a0b0-1049-461c-ae57-f8198a995144)

Maybe I can manage on my own :)

**Comment by danielkucera** [comment]
> First, I'd like to check if all of them (or most of them) follow the same spec;

I believe they all come from this reference design: https://www.semtech.com/products/wireless-rf/lora-core/sx1302cssxxxgw1#documentation

**Comment by dennis4lora** [comment]
Great, the design can support both kinds of interfaces, I think it's better than my idea to add a bridge board.

Still, I think there are compatibility issues needs to be noted.

First, I compared the miniPCIe Pins of RAK2287 and LR1302. Their SPI pins and a few other critical pins are the same; however, the pins for SX126x (used for LBT) doesn't match.

> I believe they all come from this reference design: https://www.semtech.com/products/wireless-rf/lora-core/sx1302cssxxxgw1#documentation

I then checked the Semtech reference design, Oops! totally different from RAK2287 and LR1302, as below:

![image](https://github.com/lora-gateway/esxp1302/assets/78716507/c2041da5-637b-4fa4-b52d-dbfb933bc359)

Which makes me doubt if Semtech has updated the design, while RAK2287 follows an earlier one?

So I checked a few such products selling in Chinese Market:
- [product 1](https://www.ebyte.com/product-class.html?cid=1&pid=187) follows RAK2287 design, doesn't support SX126x
- [product 2](https://item.taobao.com/item.htm?spm=a21n57.1.0.0.2c35523cJPghVV&id=679597689156&ns=1&abbucket=10#detail) follows Semtech design
- [product 3](https://item.taobao.com/item.htm?spm=a21n57.1.0.0.2c35523cJPghVV&id=652728743720&ns=1&abbucket=10#detail) follows RAK2287 design, but doesn't support SX126x

**Conclusion**:
Compatibility issue does exist. If we don't need LBT (in fact ESXP1302 doesn't support it so far), then following the RAK2287 spec should support the major miniPCIe products available from market.

**Comment by danielkucera** [comment]
> Oops! totally different from RAK2287 and LR1302

This completely slipped my attention, I was in the impression it is the same but you are right.

> following the RAK2287 spec should support the major miniPCIe products available from market.

I totally agree.

**Comment by danielkucera** [comment]
So I had it manufactured.
![IMG20240124115444](https://github.com/lora-gateway/esxp1302/assets/1734256/3458a62c-0a34-478f-ae11-549e16603977)

The biggest surprise were those 0402 resistor pads - that is really something hard to assemble at home so I just bridged them with solder.

But anyway it works as expected 👌

**Comment by dennis4lora** [comment]
Great, congratulations! :+1:

**Comment by rwijbenga** [comment]
@danielkucera Could you share the schematic for the miniPCIe wirring to the ESP32 ?

Or the board files? :)

**Comment by danielkucera** [comment]
Hi @rwijbenga ,
I don't have Altium so I just imported the board to KiCAD and botched in minipcie and wired the connections.
You can find it here:
https://github.com/danielkucera/esxp1302/tree/minipcie/hardware_design/motherboard

**Comment by M10CUBE** [comment]
Hi @danielkucera 
I came to your design working a LoraWAN solution from seeedstudio 
https://www.seeedstudio.com/WM1302-LoRaWAN-Gateway-Module-Without-SX126-SPI-EU868-p-5847.html 
I think similar to ELECROW LR1302

Our  KICAD OSHW project https://wizcube.eu project has similar PCB dimensions (90X90mm and holes 4mm from edges) a
I am thinking to make a similar design using the on board Raspberry Pico . Is that possible? I mean is any LoRaWAN software available for the RP2040 CPU? 

Since our wizcube boards have already Raspberry Pi 40 pin connector I imagine a combination of:
1 - mini PCI 1302 module from seed studio or elsewhere
https://www.seeedstudio.com/WM1302-LoRaWAN-Gateway-Module-Without-SX126-SPI-EU868-p-5847.html 
2 - A mini PCI HAT from seeedstudio or elsewhere. (may be found open source and manufacture our self from Gerber file) 
https://www.seeedstudio.com/WM1302-Pi-Hat-p-4897.html

Plugging these 2 on any wizcube.eu module we have LoRaWAN out of the box. 
So the question now is if LoRaWAN exists for Raspberry RP2040 or 2350 core? 
If not plan B is to use an ESP32-C6 CPU on a raspberry Pico footprint made by Waveshare directly plugged in on a wizcube PCB. It comes cheaper that the ESP32 DEV board. Then voila! . LoRaWAN without designing any new PCB.

I would like to have your idea since we are redesigning out wizcube.eu boards correcting some issues. 
If the above idea looks good we can include your comments to new design and of course everybody can benefit from our OSHW concept

I could like to have your opinion and help me on taking the right decision.
Thanks 
Vasilis

**Comment by danielkucera** [comment]
Hi @M10CUBE ,
LoRaWAN most likely exists for RP2040 but not with SX13xx drivers, just as client.

**Comment by dennis4lora** [comment]
I agree with @danielkucera that there is no suitable sx130x project for RP2040 yet. Search on Internet, I find some results using RP2040 with LoRa as **end device**, and I'm afraid they are only based on the drivers for Rx and Tx, not even compatible with LoRaWAN protocol.

Currently, there are two LoRaWAN implementations provided by Semtech for **end device**:
- [LoRaMAC-Node](https://github.com/Lora-net/loramac-node) is deprecated
- [LBM](https://github.com/Lora-net/SWL2001) has been ported to [NRF52840](https://github.com/Lora-net/SWL2001/tree/master/lbm_applications/2_porting_nrf_52840)

but none works on Raspberry Pico.

OK, back to the **gateway** topic. 😄

This ESXP1302 project is tightly bound with ESP32 since it based on [esp-idf](https://github.com/espressif/esp-idf) using lots of services from that framework, and it also uses the Wi-Fi from ESP32 as its backhaul. Which means you can't replace ESP32 with another MCU and expect the firmware still works.

But if you only use ESP32-C6 to replace the original ESP32 MCU, I expect there would just need very trivial changes to make it working. You may find #23 is interesting.

**Comment by M10CUBE** [comment]
Thanks ,
> Currently, there are two LoRaWAN implementations provided by Semtech for **end device**:
> 
>     * [LoRaMAC-Node](https://github.com/Lora-net/loramac-node) is deprecated
> 
>     * [LBM](https://github.com/Lora-net/SWL2001) has been ported to [NRF52840](https://github.com/Lora-net/SWL2001/tree/master/lbm_applications/2_porting_nrf_52840)
> 
> but none works on Raspberry Pico.
> 
OK time to forget RP2040 at the moment as being LoRaWAN device. Lets move on
> OK, back to the **gateway** topic. 😄
> This ESXP1302 project is tightly bound with ESP32 since it based on [esp-idf](https://github.com/espressif/esp-idf) using lots of services from that framework, and it also uses the Wi-Fi from ESP32 as its backhaul. Which means you can't replace ESP32 with another MCU and expect the firmware still works.
You may find #23 is interesting.
>
The #23 is fantastic. I will get one. I have XIAO ESP32-C6 in the LAB.

> But if you only use ESP32-C6 to replace the original ESP32 MCU, I expect there would just need very trivial changes to make it working. 
>
I  absolutely agree and the solution looks one way road:

To try the above since it does not cost me any KiCAD time. Maybe some simple wiring under the board
I also asked here and apply it to project ARGO (nRF9151) .
https://github.com/craigpeacock/Zephyr_LoRaWAN/issues/2#issue-2477758665

I think Zephyr is the keyword. 
If the chip supports Zephyr you have tons of drivers out of the box. That is the case with Nordic chips.
As you Know Arduino is moving to Zephyr now.

**Comment by dennis4lora** [comment]
@HeadBoffin:

You are right, thanks for pointing out those 2 projects, which are both work on Raspberry Pico as LoRaWAN client.

As for Arduino, I guess it doesn't matter whether it uses Zephyr not. And for ESXP1302, it's based on `esp-idf` which bases on FreeRTOS, so it's hard to switch to Arduino or Zephyr...

**Comment by dennis4lora** [comment]
@HeadBoffin:

You are right, thanks for pointing out those 2 projects, which are both work on Raspberry Pico as LoRaWAN client.

As for Arduino, I guess it doesn't matter whether it uses Zephyr not. And for ESXP1302, it's based on `esp-idf` which bases on FreeRTOS, so it's hard to switch to Arduino or Zephyr...

**Comment by PureTek-Innovations** [comment]
Lilygo now sells an ESP32 with an optional PCIe adapter
https://www.lilygo.cc/products/t-eth-elite-1?variant=44498204983477

They also have a very similar codebase with working ethernet
https://github.com/Xinyuan-LilyGO/LilyGO-ETH-Gateway

I've ordered a set to try.

**Comment by danielkucera** [comment]
They even have the same typos in the code:
```
static int dns_loopup
```

**Comment by edil-project** [comment]
> Olá,
> 
> Estou usando este firmware com sucesso com o [LR1302 da Elecrow](https://www.elecrow.com/lr1302-lorawan-gateway-module-spi-eu868-sx1302-long-range-gateway-module-support-8-channels.html) e minha [placa de breakout miniPCIe](https://lectronz.com/products/minipci-e-breakout-board) . Esta é minha configuração atual.
> 
> ![IMG20231217152824](https://github.com/lora-gateway/esxp1302/assets/1734256/a637d0cd-d4ff-4bd9-8c4b-d3732de7dabf)
> 
> Você consideraria adicionar um conector miniPCIe ao seu projeto de hardware atual, de acordo com [essa especificação](https://www.elecrow.com/wiki/index.php?title=LR1302_LoRaWAN_Gateway_Module#Interface_Function) ? Testei a mesma configuração também para [o RAK2287](https://docs.rakwireless.com/Product-Categories/WisLink/RAK2287/Datasheet/#pinout-description) e funcionou.
> 
> Eu estava pensando em fazer isso eu mesmo, mas não tenho o software de design para editar seu projeto e estou com preguiça de criar um do zero.

Olá, belo projeto! Qual esquema para os pinos e firmware que usou?

**Comment by danielkucera** [comment]
This is my latest prototype BTW. It has been in "production" for several months.

![Image](https://github.com/user-attachments/assets/d95b5617-5caf-4e5f-b641-7c5903e88856)

**Comment by edil-project** [comment]
> Aliás, este é o meu protótipo mais recente. Ele está em "produção" há vários meses.
> 
> ![Image](https://github.com/user-attachments/assets/d95b5617-5caf-4e5f-b641-7c5903e88856)

poderia compartilhar o esquema de ligação da pci?
Vou iniciar um projeto!
